### PR TITLE
Implement the price subscription through trigger/redis

### PR DESCRIPTION
**Pull request by samerbuna**

### CODE REVIEW

**Comment by bodymindarts** [comment]
I'm not sure we should increase the dependency on Redis here. Using a pub-sub system for this use-case is certainly the right way to go but I doubt Redis is the right system for pub-sub in the future.
Perhaps its no big deal to switch implementation later... need to discuss

**Review by bodymindarts — DISMISSED** [review]

File: test/jest-unit.config.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ module.exports = {
   },
   testRegex: ".*\\.spec\\.ts$",
   testSequencer: "<rootDir>/test/jest-test-sequencer.js",
+  setupFilesAfterEnv: ["<rootDir>/test/jest.setup.js"],
```
Comment (dolcalmi) [inline test/jest-unit.config.js:10]: We should not have this in unit testing, if this is needed then is something wrong with code/dependencies

**Review by dolcalmi — CHANGES_REQUESTED** [review]

File: test/jest-unit.config.js (hunk 1 above)
Comment (samerbuna) [reply test/jest-unit.config.js]: Fixed

File: src/graphql/root/subscription/price.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        }),
+      )
+      return pubsub.asyncIterator(eventName)
+    }
+
+    const satUsdPrice = await getCurrentPrice()
+
+    if (satUsdPrice) {
+      setTimeout(
+        () => pubsub.publish(eventName, { satUsdCentPrice: 100 * satUsdPrice }),
```
Comment (dolcalmi) [inline src/graphql/root/subscription/price.ts:61]: not sure if I am not undestanding but `satUsdCentPrice` should be `price`? (like in trigger)

File: src/graphql/root/subscription/price.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  subscribe: async (_, args) => {
+    const { amountCurrencyUnit, priceCurrencyUnit } = args.input
+
+    for (const input of [amountCurrencyUnit, priceCurrencyUnit]) {
+      if (input instanceof Error) {
+        return { errors: [{ message: input.message }] }
+      }
+    }
+
+    const eventName = "SAT-USDCENT-PRICE"
```
Comment (dolcalmi) [inline src/graphql/root/subscription/price.ts:45]: maybe this should be in a common file/config (to be used in trigger too)

File: src/graphql/types/scalar/big-int.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  const num = Number(value)
+  if (num !== num || num > MAX_INT || num < MIN_INT) {
+    throw new Error(
+      "BigInt cannot represent non 53-bit signed integer value: " + String(value),
+    )
+  }
+  const int = Math.floor(num)
+  if (int !== num) {
+    throw new Error("BigInt cannot represent non-integer value: " + String(value))
+  }
```
Comment (dolcalmi) [inline src/graphql/types/scalar/big-int.ts:39]: the subscription is working but after a couple of seconds I got "message": "BigInt cannot represent non 53-bit signed integer value: NaN" or maybe is this `num !== num`

price 0.000465380546875

**Review by dolcalmi — CHANGES_REQUESTED** [review]

File: src/graphql/root/subscription/price.ts (hunk 3 above)
Comment (samerbuna) [reply src/graphql/root/subscription/price.ts]: That's actually what's causing the NaN error. Fixed.

File: src/graphql/types/scalar/big-int.ts (hunk 4 above)
Comment (samerbuna) [reply src/graphql/types/scalar/big-int.ts:39]: Fixed

File: src/graphql/root/subscription/price.ts (hunk 2 above)
Comment (samerbuna) [reply src/graphql/root/subscription/price.ts]: That's actually what's causing the NaN error. Fixed.

File: src/graphql/types/scalar/big-int.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import { GT } from "@graphql/index"
+
+const MAX_INT = Number.MAX_SAFE_INTEGER
+const MIN_INT = Number.MIN_SAFE_INTEGER
+
+const BigInt = new GT.Scalar({
```
[the remark this answers is not in the document]
Comment (samerbuna) [reply src/graphql/types/scalar/big-int.ts]: Renamed

File: src/graphql/root/subscription/price.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      return { errors: source.errors }
+    }
+    const amountPriceInCents = args.input.amount * source.satUsdCentPrice
+    return {
+      errors: [],
+      price: {
+        formattedAmount: amountPriceInCents.toString(),
+        base: Math.round(amountPriceInCents * 10 ** 16),
+        offset: 16,
+        currencyUnit: "USDCENT",
```
[the remark this answers is not in the document]
Comment (samerbuna) [reply src/graphql/root/subscription/price.ts:33]: I've seen prices where we need 12 or more decimal offset to be accurate

File: src/graphql/types/scalar/safe-int.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+})
+
+function coerceSafeInt(value) {
+  if (value === "") {
+    throw new Error(
+      "SafeInt cannot represent non 53-bit signed integer value: (empty string)",
+    )
+  }
+  const num = Number(value)
+  if (num !== num || num > MAX_INT || num < MIN_INT) {
```
Comment (dolcalmi) [inline src/graphql/types/scalar/safe-int.ts:31]: `num !== num` should be `num !== value`

File: src/graphql/root/subscription/price.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        }),
+      )
+      return pubsub.asyncIterator(eventName)
+    }
+
+    const satUsdPrice = await getCurrentPrice()
+
+    if (satUsdPrice) {
+      setTimeout(
+        () => pubsub.publish(eventName, { satUsdPrice: 100 * satUsdPrice }),
```
Comment (dolcalmi) [inline src/graphql/root/subscription/price.ts:62]: this should use `satUsdCentPrice` instead of  `satUsdPrice`

**Review by dolcalmi — CHANGES_REQUESTED** [review]

File: src/graphql/root/subscription/price.ts (hunk 8 above)
Comment (samerbuna) [reply src/graphql/root/subscription/price.ts]: Fixed

File: src/graphql/types/scalar/safe-int.ts (hunk 7 above)
Comment (samerbuna) [reply src/graphql/types/scalar/safe-int.ts]: Changed

**Review by dolcalmi — APPROVED** [review]

**Review by vindard — APPROVED** [review]
No new issues found past the already-resolved comments on here from my end, LGTM 🎉

**Pull request merged**